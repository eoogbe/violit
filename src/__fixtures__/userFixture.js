/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const username = 'fred-astaire';

export const password = 'password123';

export const credentials = {
  username,
};

export const authUrl = 'http://localhost:4000/api/auth';

export const usersUrl = 'http://localhost:4000/api/users';

export const userUrl = 'http://localhost:4000/api/users/fred-astaire';

export const authApi = [
  {
    href: authUrl,
    data: [
      { name: 'username', value: username },
    ],
  }
];

export const userApi = [
  {
    href: userUrl,
    data: [
      { name: 'username', value: username },
    ],
  }
];

export const credentialsFetch = {
  url: authUrl,
  method: 'GET',
  status: 200,
  body: {
    items: userApi,
  },
};

export const loginFetch = {
  url: authUrl,
  method: 'POST',
  status: 201,
  body: {
    items: userApi,
  },
};

export const logoutFetch = {
  url: authUrl,
  method: 'DELETE',
  status: 200,
};

export const signupFetch = {
  url: usersUrl,
  method: 'POST',
  status: 201,
  body: {
    items: userApi,
  },
};

export const hasUserFetch = {
  url: userUrl,
  method: 'HEAD',
  status: 200,
};

export const doesNotHaveUserFetch = {
  url: userUrl,
  method: 'HEAD',
  status: 404,
};
