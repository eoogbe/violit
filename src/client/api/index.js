/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
