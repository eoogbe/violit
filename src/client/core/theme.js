/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default {
  primaryColor: '#7F00FF',

  gray200: '#EEEEEE',
  gray300: '#E0E0E0',
  gray500: '#9E9E9E',
  gray700: '#616161',

  red200: '#EF9A9A',
  red500: '#F44336',

  lightBlue700: '#0288D1',

  black: '#000000',
  white: '#FFFFFF',

  spacingSm: '0.25em',
  spacingMd: '0.5em',
  spacingLg: '1em',

  borderRadius: '2px',
};
