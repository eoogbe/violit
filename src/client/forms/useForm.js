/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import camelCase from 'camel-case';
import * as R from 'ramda';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/** Hook that adds state for a form with controlled components. */
export default (initialValues, validate, options) => {
  const { commitOnChange } = options || {};

  const prevValues = useRef(initialValues);
  const preventDoubleSubmit = useRef(false);
  const [values, setValues] = useState(initialValues);
  const [errors, setErrors] = useState({});
  const [dirtied, setDirtied] = useState(R.map(R.F, initialValues));
  const [submitting, setSubmitting] = useState(false);

  const unmounted = useRef(false);
  useEffect(() => () => {
    unmounted.current = true;
  }, []);

  const isDirty = useMemo(() => {
    return Object.values(dirtied).includes(true);
  }, [dirtied]);

  const hasError = useMemo(() => {
    return Object.values(errors).some((error) => !!error);
  }, [errors]);

  const validateFields = useCallback(async () => {
    if (!validate) {
      return {};
    }

    const errors = await validate(values) || {};
    if (!unmounted.current) {
      setErrors(errors);
    }

    return errors;
  }, [validate, values]);

  useEffect(() => {
    if (values === prevValues.current) {
      return;
    }

    validateFields();
    prevValues.current = values;
  }, [values, validateFields]);

  const reset = useCallback(() => {
    if (unmounted.current) {
      return;
    }

    prevValues.current = initialValues;
    preventDoubleSubmit.current = false;
    setValues(initialValues);
    setErrors({});
    setDirtied(R.map(R.F, initialValues));
    setSubmitting(false);
  }, [initialValues]);

  const handleInput = useCallback(({ target }) => {
    const { name, value } = target;
    setValues(R.assoc(camelCase(name), value));
  }, []);

  const handleCommit = useCallback(({ target }) => {
    const { value } = target;
    const name = camelCase(target.name);
    setValues(R.assoc(name, value));
    setDirtied(R.assoc(name, true));
  }, []);

  const handleChange = useCallback((event) => {
    handleInput(event);
    if (commitOnChange) {
      handleCommit(event);
    }
  }, [handleInput, handleCommit, commitOnChange]);

  const handleSubmit = useCallback((onSubmit) => async (event) => {
    event.preventDefault();

    if (preventDoubleSubmit.current) {
      return;
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = true;
      setSubmitting(true);
    }

    const fieldErrors = await validateFields();
    if (Object.values(fieldErrors).every((error) => !error)) {
      const submitErrors = await onSubmit(values) || {};
      if (!unmounted.current) {
        setErrors(submitErrors);
      }
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = false;
      setSubmitting(false);
    }
  }, [validateFields, values]);

  const eventHandlers = {
    handleChange,
    handleBlur: handleCommit,
    handleSubmit,
  };

  return [
    values,
    errors,
    dirtied,
    eventHandlers,
    { submitting, isDirty, hasError, reset, validateFields },
  ];
};
