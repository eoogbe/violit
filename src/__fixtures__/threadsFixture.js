/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as USER from './userFixture';

export const thread1 = {
  id: 'thread1',
  slug: 'lorem-ipsum',
  dateCreated: '2019-09-29T01:23:45.679Z',
  dateDeleted: '2019-09-29T02:00:45.679Z',
  headline: 'Lorem ipsum',
  articleBody: 'Dolor sit amet, consectetur adipiscing elit.',
  author: USER.username,
  title: 'Lorem ipsum | Violit',
  path: '/threads/lorem-ipsum',
  url: 'http://localhost:4000/api/threads/lorem-ipsum',
  deleteUrl: 'http://localhost:4000/api/threads/lorem-ipsum/delete',
  commentsUrl: 'http://localhost:4000/api/threads/lorem-ipsum/comments',
};

export const thread2 = {
  id: 'thread2',
  slug: 'aliquip-ex-ea',
  dateCreated: '2019-09-28T09:08:07.654Z',
  dateDeleted: '2019-09-28T09:08:08.654Z',
  headline: 'Aliquip ex ea',
  articleBody: null,
  author: USER.username,
  title: 'Aliquip ex ea | Violit',
  path: '/threads/aliquip-ex-ea',
  url: 'http://localhost:4000/api/threads/aliquip-ex-ea',
  deleteUrl: 'http://localhost:4000/api/threads/aliquip-ex-ea/delete',
  commentsUrl: 'http://localhost:4000/api/threads/aliquip-ex-ea/comments',
};

export const thread3 = {
  id: 'thread3',
  slug: 'commodo-consequat',
  dateCreated: '2019-09-27T03:21:09.876Z',
  dateDeleted: '2019-09-27T04:21:09.876Z',
  headline: 'Commodo consequat',
  articleBody: null,
  author: USER.username,
  title: 'Commodo consequat | Violit',
  path: '/threads/commodo-consequat',
  url: 'http://localhost:4000/api/threads/commodo-consequat',
  deleteUrl: 'http://localhost:4000/api/threads/commodo-consequat/delete',
  commentsUrl: 'http://localhost:4000/api/threads/commodo-consequat/comments',
};

export const createInput = {
  headline: thread1.headline,
  articleBody: thread1.articleBody,
};

export const normalized = {
  thread1: {
    id: 'thread1',
    slug: thread1.slug,
    dateCreated: thread1.dateCreated,
    dateDeleted: null,
    headline: thread1.headline,
    articleBody: thread1.articleBody,
    author: thread1.author,
  },
  thread2: {
    id: 'thread2',
    slug: thread2.slug,
    dateCreated: thread2.dateCreated,
    dateDeleted: thread2.dateDeleted,
    headline: thread2.headline,
    articleBody: thread2.articleBody,
    author: thread2.author,
  },
  thread3: {
    id: 'thread3',
    slug: thread3.slug,
    dateCreated: thread3.dateCreated,
    dateDeleted: null,
    headline: thread3.headline,
    articleBody: thread3.articleBody,
    author: thread3.author,
  },
};

export const unnormalized = [
  {
    id: 'thread1',
    slug: thread1.slug,
    dateCreated: thread1.dateCreated,
    headline: thread1.headline,
    articleBody: thread1.articleBody,
    author: thread1.author,
  },
  {
    id: 'thread3',
    slug: thread3.slug,
    dateCreated: thread3.dateCreated,
    headline: thread3.headline,
    articleBody: thread3.articleBody,
    author: thread3.author,
  },
];

export const api = [
  {
    href: thread1.url,
    data: [
      { name: 'id', value: 'thread1' },
      { name: 'slug', value: thread1.slug },
      { name: 'dateCreated', value: thread1.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'headline', value: thread1.headline },
      { name: 'articleBody', value: thread1.articleBody },
      { name: 'author', value: thread1.author },
    ],
    links: [
      { rel: 'deleteAction', href: thread1.deleteUrl },
      { rel: 'isPartOf', href: 'http://localhost:4000/api/board' },
      {
        rel: 'comment',
        href: thread1.commentsUrl,
      },
    ],
  },
  {
    href: thread3.url,
    data: [
      { name: 'id', value: 'thread3' },
      { name: 'slug', value: thread3.slug },
      { name: 'dateCreated', value: thread3.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'headline', value: thread3.headline },
      { name: 'articleBody', value: '' },
      { name: 'author', value: thread3.author },
    ],
    links: [
      { rel: 'deleteAction', href: thread3.deleteUrl },
      { rel: 'isPartOf', href: 'http://localhost:4000/api/board' },
      {
        rel: 'comment',
        href: thread3.commentsUrl,
      },
    ],
  },
];

export const deleteThreadApi = [
  {
    href: thread1.deleteUrl,
    data: [
      { name: 'actionStatus', value: 'completedActionStatus' },
      { name: 'endTime', value: thread1.dateDeleted },
    ],
    links: [
      { rel: 'object', href: thread1.url },
      { rel: 'targetCollection', href: 'http://localhost:4000/api/board' },
    ],
  },
];

export const threadsFetch = {
  url: 'http://localhost:4000/api/threads',
  method: 'GET',
  status: 200,
  body: {
    items: api,
  },
};

export const threadFetch1 = {
  url: thread1.url,
  method: 'GET',
  status: 200,
  body: {
    items: [api[0]],
  },
};

export const threadWithoutArticleBodyFetch = {
  url: thread1.url,
  method: 'GET',
  status: 200,
  body: {
    items: [
      {
        href: thread1.url,
        data: [
          { name: 'id', value: 'thread1' },
          { name: 'slug', value: thread1.slug },
          { name: 'dateCreated', value: thread1.dateCreated },
          { name: 'dateDeleted', value: '' },
          { name: 'headline', value: thread1.headline },
          { name: 'articleBody', value: '' },
          { name: 'author', value: thread1.author },
        ],
        links: [
          { rel: 'deleteAction', href: thread1.deleteUrl },
          { rel: 'isPartOf', href: 'http://localhost:4000/api/board' },
          {
            rel: 'comment',
            href: thread1.commentsUrl,
          },
        ],
      },
    ],
  },
};

export const threadWhereCannotCreateCommentFetch = {
  url: thread1.url,
  method: 'GET',
  status: 200,
  body: {
    items: [
      {
        href: thread1.url,
        data: [
          { name: 'id', value: 'thread1' },
          { name: 'slug', value: thread1.slug },
          { name: 'dateCreated', value: thread1.dateCreated },
          { name: 'dateDeleted', value: '' },
          { name: 'headline', value: thread1.headline },
          { name: 'articleBody', value: thread1.articleBody },
          { name: 'author', value: thread1.author },
        ],
        links: [
          { rel: 'isPartOf', href: 'http://localhost:4000/api/board' },
          {
            rel: 'comment',
            href: thread1.commentsUrl,
          },
        ],
      },
    ],
  },
};

export const deletedThreadFetch = {
  url: thread1.url,
  method: 'GET',
  status: 200,
  body: {
    items: [
      {
        href: thread1.url,
        data: [
          { name: 'id', value: 'thread1' },
          { name: 'slug', value: thread1.slug },
          { name: 'dateCreated', value: thread1.dateCreated },
          { name: 'dateDeleted', value: thread1.dateDeleted },
          { name: 'headline', value: thread1.headline },
          { name: 'articleBody', value: '' },
          { name: 'author', value: thread1.author },
        ],
        links: [
          { rel: 'isPartOf', href: 'http://localhost:4000/api/board' },
          {
            rel: 'comment',
            href: thread1.commentsUrl,
          },
        ],
      },
    ],
  },
};

export const createThreadFetch = {
  url: 'http://localhost:4000/api/threads',
  method: 'POST',
  status: 201,
  body: {
    items: [api[0]],
  },
};

export const canCreateThreadFetch = {
  url: 'http://localhost:4000/api/threads',
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS,POST',
  },
};

export const cannotCreateThreadFetch = {
  url: 'http://localhost:4000/api/threads',
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS',
  },
};

export const deleteThreadFetch = {
  url: thread1.deleteUrl,
  method: 'POST',
  status: 200,
  body: {
    items: deleteThreadApi,
  },
};

export const db = [
  {
    id: 'thread1',
    slug: thread1.slug,
    dateCreated: thread1.dateCreated,
    dateDeleted: null,
    headline: thread1.headline,
    articleBody: thread1.articleBody,
    userId: 'user1',
    boardId: 'board1',
  },
  {
    id: 'thread2',
    slug: thread2.slug,
    dateCreated: thread2.dateCreated,
    dateDeleted: thread2.dateDeleted,
    headline: thread2.headline,
    articleBody: thread2.articleBody,
    userId: 'user1',
    boardId: 'board1',
  },
  {
    id: 'thread3',
    slug: thread3.slug,
    dateCreated: thread3.dateCreated,
    dateDeleted: null,
    headline: thread3.headline,
    articleBody: thread3.articleBody,
    userId: 'user1',
    boardId: 'board1',
  },
];
