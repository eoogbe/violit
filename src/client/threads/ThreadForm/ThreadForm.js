/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import { Redirect, useHistory } from 'react-router-dom';
import styled from 'styled-components';
import { Alert, Button, useTitle } from 'client/core';
import { FormError, Input, useForm } from 'client/forms';
import { useLoad } from 'client/loading';
import * as api from '../api';

const StyledThreadForm = styled.form`
  font-size: 1rem;
  display: flex;
  flex-direction: column;
  padding: ${({ theme }) => theme.spacingMd};
`;

const Heading = styled.h2`
  font-size: 1.75rem;
  margin: 0;
  margin-bottom: 0.288em;
`;

const FieldList = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`;

FieldList.displayName = 'FieldList';

const SubmitButton = styled(Button)`
  align-self: flex-end;
`;

/** The screen for creating a thread. */
const ThreadForm = () => {
  const history = useHistory();
  const [unknownError, setUnknownError] = useState(false);

  const [
    loading,
    loadError,
    canCreateThread,
  ] = useLoad(api.checkCanCreateThread);

  useTitle('New thread');

  const [
    values,
    errors,
    dirtied,
    { handleChange, handleSubmit },
    { submitting, isDirty, hasError: formError },
  ] = useForm({
    headline: '',
    articleBody: '',
  }, ({ headline }) => {
    setUnknownError(false);
    return {
      headline: !headline.trimEnd() && 'Must not be blank.',
    };
  }, {
    commitOnChange: true,
  });
  const disabled = !isDirty || formError || submitting;

  const onSubmit = async (input) => {
    setUnknownError(false);
    try {
      const thread = await api.sendCreateThread(input);
      if (thread.error) {
        return {
          form: thread.error,
        };
      } else {
        history.replace(`/threads/${thread.slug}`);
      }
    } catch (err) {
      setUnknownError(true);
    }
  };

  const showAlert = loadError || unknownError;

  if (!loading && !loadError && !canCreateThread) {
    return <Redirect to="/login" />;
  }

  return (
    <StyledThreadForm onSubmit={handleSubmit(onSubmit)}>
      <Heading>New thread</Heading>
      {errors.form && <FormError>{errors.form}</FormError>}
      <FieldList>
        <li>
          <label htmlFor="headline">Title</label>
          <Input
            id="headline"
            type="text"
            name="headline"
            aria-required="true"
            value={values.headline}
            onChange={handleChange}
            autoFocus
          />
          {dirtied.headline && errors.headline &&
            <FormError aria-describedby="headline">
              {errors.headline}
            </FormError>}
        </li>
        <li>
          <label htmlFor="article-body">Body</label>
          <Input
            as="textarea"
            id="article-body"
            name="article-body"
            rows={5}
            value={values.articleBody}
            onChange={handleChange}
          />
        </li>
      </FieldList>
      <SubmitButton type="submit" disabled={disabled} loading={submitting}>
        Post
      </SubmitButton>
      {showAlert && <Alert />}
    </StyledThreadForm>
  );
};

export default ThreadForm;
