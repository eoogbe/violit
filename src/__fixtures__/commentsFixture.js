/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as THREADS from './threadsFixture';
import * as USER from './userFixture';

export const comment1 = {
  id: 'comment1',
  dateCreated: '2019-10-09T10:11:12.123Z',
  dateDeleted: '2019-10-09T12:11:12.123Z',
  text: 'Sed do eiusmod tempor incididunt ut labore et dolore.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const comment2 = {
  id: 'comment2',
  dateCreated: '2019-10-08T07:08:09.101Z',
  dateDeleted: '2019-10-08T07:09:09.101Z',
  text: 'Ut enim ad minim veniam.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const comment3 = {
  id: 'comment3',
  dateCreated: '2019-09-30T01:02:03.456Z',
  dateDeleted: '2019-10-03T01:02:03.456Z',
  text: 'Quis nostrud exercitation ullamco laboris nisi ut.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const createInput = {
  id: 'comment1',
  text: comment1.text,
  threadSlug: THREADS.thread1.slug,
};

export const normalized = {
  comment1: {
    id: 'comment1',
    dateCreated: comment1.dateCreated,
    dateDeleted: null,
    text: comment1.text,
    threadSlug: comment1.threadSlug,
    author: comment1.author,
  },
  comment2: {
    id: 'comment2',
    dateCreated: comment2.dateCreated,
    dateDeleted: comment2.dateDeleted,
    text: comment2.text,
    threadSlug: comment2.threadSlug,
    author: comment2.author,
  },
  comment3: {
    id: 'comment3',
    dateCreated: comment3.dateCreated,
    dateDeleted: null,
    text: comment3.text,
    threadSlug: comment3.threadSlug,
    author: comment3.author,
  },
};

export const unnormalized = [
  normalized.comment1,
  normalized.comment2,
  normalized.comment3,
];

export const api = [
  {
    href: 'http://localhost:4000/api/comments/comment1',
    data: [
      { name: 'id', value: 'comment1' },
      { name: 'dateCreated', value: comment1.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'text', value: comment1.text },
      { name: 'author', value: comment1.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      {
        rel: 'deleteAction',
        href: 'http://localhost:4000/api/comments/comment1/delete',
      },
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
  {
    href: 'http://localhost:4000/api/comments/comment2',
    data: [
      { name: 'id', value: 'comment2' },
      { name: 'dateCreated', value: comment2.dateCreated },
      { name: 'dateDeleted', value: comment2.dateDeleted },
      { name: 'text', value: comment2.text },
      { name: 'author', value: comment2.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
  {
    href: 'http://localhost:4000/api/comments/comment3',
    data: [
      { name: 'id', value: 'comment3' },
      { name: 'dateCreated', value: comment3.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'text', value: comment3.text },
      { name: 'author', value: comment3.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      {
        rel: 'deleteAction',
        href: 'http://localhost:4000/api/comments/comment3/delete',
      },
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
];

export const deleteCommentApi = [
  {
    href: 'http://localhost:4000/api/comments/comment1/delete',
    data: [
      { name: 'actionStatus', value: 'completedActionStatus' },
      { name: 'endTime', value: comment1.dateDeleted },
    ],
    links: [
      {
        rel: 'object',
        href: 'http://localhost:4000/api/comments/comment1',
      },
      { rel: 'targetCollection', href: THREADS.thread1.commentsUrl },
    ],
  },
];

export const commentsFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'GET',
  status: 200,
  body: {
    items: api,
  },
};

export const createCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'POST',
  status: 201,
  body: {
    items: [api[0]],
  },
};

export const canCreateCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS,POST',
  },
};

export const cannotCreateCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS',
  },
};

export const deleteCommentFetch = {
  url: 'http://localhost:4000/api/comments/comment1/delete',
  method: 'POST',
  status: 200,
  body: {
    items: deleteCommentApi,
  },
};

export const db = [
  {
    id: 'comment1',
    dateCreated: comment1.dateCreated,
    dateDeleted: null,
    content: comment1.text,
    userId: 'user1',
    threadId: 'thread1',
  },
  {
    id: 'comment2',
    dateCreated: comment2.dateCreated,
    dateDeleted: comment2.dateDeleted,
    content: comment2.text,
    userId: 'user1',
    threadId: 'thread1',
  },
  {
    id: 'comment3',
    dateCreated: comment3.dateCreated,
    dateDeleted: null,
    content: comment3.text,
    userId: 'user1',
    threadId: 'thread1',
  },
];
