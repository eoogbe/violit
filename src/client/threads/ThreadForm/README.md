```jsx
import { MemoryRouter } from 'react-router-dom';
import * as THREADS from '__fixtures__/threadsFixture';

mockFetch([
  THREADS.canCreateThreadFetch,
]);

<MemoryRouter initialEntries={['/new-thread']}>
  <ThreadForm />
</MemoryRouter>
```
