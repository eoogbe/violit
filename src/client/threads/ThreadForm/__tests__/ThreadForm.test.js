/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import * as ReactRouterDom from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import * as THREADS from '__fixtures__/threadsFixture';
import { Alert, Button } from 'client/core';
import { FormError, Input } from 'client/forms';
import mockFetch from 'testUtils/mockFetch';
import { updateWrapper, wait } from 'testUtils/wait';
import ThreadForm from '../ThreadForm';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(() => ({
    replace: () => {},
  })),
}));

const changeHeadline = (wrapper, value) => {
  wrapper.find(Input).first().simulate('change', {
    target: {
      name: 'headline',
      value,
    },
  });
};

const changeArticleBody = (wrapper, value) => {
  wrapper.find(Input).at(1).simulate('change', {
    target: {
      name: 'article-body',
      value,
    },
  });
};

const simulateSubmit = async (wrapper, waitFn = wait) => {
  const onSubmit = wrapper.find(ThreadForm).children().prop('onSubmit');
  return act(async () => {
    await onSubmit({
      preventDefault: () => {},
    });
    await waitFn();
  });
};

const getFieldErrorWrapper = (wrapper, index) =>
  wrapper
    .find('FieldList')
    .children()
    .children()
    .children()
    .at(index)
    .find(FormError);

it('renders without crashing', async () => {
  mockFetch([
    THREADS.canCreateThreadFetch,
  ]);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <ReactRouterDom.MemoryRouter>
        <ThreadForm />
      </ReactRouterDom.MemoryRouter>
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('sets the document title', async () => {
  mockFetch([
    THREADS.canCreateThreadFetch,
  ]);

  await TestRenderer.act(async () => {
    TestRenderer.create(
      <ReactRouterDom.MemoryRouter>
        <ThreadForm />
      </ReactRouterDom.MemoryRouter>
    );

    await wait();
  });

  expect(document.title).toContain('New thread');
});

it('when cannot create thread redirects to the login screen', async () => {
  mockFetch([
    THREADS.cannotCreateThreadFetch,
  ]);

  let location;
  await TestRenderer.act(async () => {
    TestRenderer.create(
      <ReactRouterDom.MemoryRouter>
        <ThreadForm />
        <ReactRouterDom.Route
          path="*"
          render={(props) => {
            location = props.location;
            return null;
          }}
        />
      </ReactRouterDom.MemoryRouter>
    );

    await wait();
  });

  expect(location).toHaveProperty('pathname', '/login');
});

it('when submitting redirects to the new thread location', async () => {
  const historyReplace = jest.fn();
  ReactRouterDom.useHistory.mockReturnValue({
    replace: historyReplace,
  });

  mockFetch([
    THREADS.canCreateThreadFetch,
    THREADS.createThreadFetch,
  ]);

  const wrapper = mount(
    <ReactRouterDom.MemoryRouter>
      <ThreadForm />
    </ReactRouterDom.MemoryRouter>
  );
  expect(wrapper.find(Button).prop('disabled')).toBe(true);

  changeHeadline(wrapper, THREADS.thread1.headline);
  expect(wrapper.find(Button).prop('disabled')).toBe(false);

  changeArticleBody(wrapper, THREADS.thread1.articleBody);

  await simulateSubmit(wrapper, () => updateWrapper(wrapper));

  expect(historyReplace).toHaveBeenCalledWith(THREADS.thread1.path);
});

it('with submit error sets form error', async () => {
  mockFetch([
    THREADS.canCreateThreadFetch,
    {
      url: 'http://localhost:4000/api/threads',
      method: 'POST',
      status: 422,
      body: {
        error: {
          message: 'Headline is required.',
        },
      },
    },
  ]);

  const wrapper = mount(
    <ReactRouterDom.MemoryRouter>
      <ThreadForm />
    </ReactRouterDom.MemoryRouter>
  );
  expect(wrapper.find(FormError).first()).toHaveLength(0);

  changeHeadline(wrapper, THREADS.thread1.headline);

  await simulateSubmit(wrapper, () => updateWrapper(wrapper));

  expect(wrapper.find(FormError).first().text()).toBe('Headline is required.');
  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});

it('with unknown error renders the alert', async () => {
  mockFetch([
    THREADS.canCreateThreadFetch,
    {
      url: 'http://localhost:4000/api/threads',
      method: 'POST',
      status: 500,
    },
  ]);

  const wrapper = mount(
    <ReactRouterDom.MemoryRouter>
      <ThreadForm />
    </ReactRouterDom.MemoryRouter>
  );
  expect(wrapper.find(Alert)).toHaveLength(0);

  changeHeadline(wrapper, THREADS.thread1.headline);

  await simulateSubmit(wrapper, () => updateWrapper(wrapper));

  expect(wrapper.find(Alert)).toHaveLength(1);
});

it('with blank headline sets field error', async () => {
  mockFetch([
    THREADS.canCreateThreadFetch,
  ]);

  const wrapper = mount(
    <ReactRouterDom.MemoryRouter>
      <ThreadForm />
    </ReactRouterDom.MemoryRouter>
  );

  let errorWrapper = getFieldErrorWrapper(wrapper, 0);
  expect(errorWrapper).toHaveLength(0);

  changeHeadline(wrapper, '  ');

  await updateWrapper(wrapper);

  errorWrapper = getFieldErrorWrapper(wrapper, 0);
  expect(errorWrapper.text()).toBe('Must not be blank.');

  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});
