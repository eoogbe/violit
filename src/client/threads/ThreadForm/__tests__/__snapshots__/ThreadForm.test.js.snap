// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<form
  className="sc-hMqMXs dEZLlH"
  onSubmit={[Function]}
>
  <h2
    className="sc-kEYyzF jZYaHt"
  >
    New thread
  </h2>
  <ul
    className="sc-kkGfuU tdPwr"
  >
    <li>
      <label
        htmlFor="headline"
      >
        Title
      </label>
      <input
        aria-required="true"
        autoFocus={true}
        className="sc-htoDjs kEmxlR"
        id="headline"
        name="headline"
        onChange={[Function]}
        type="text"
        value=""
      />
    </li>
    <li>
      <label
        htmlFor="article-body"
      >
        Body
      </label>
      <textarea
        className="sc-htoDjs kEmxlR"
        id="article-body"
        name="article-body"
        onChange={[Function]}
        rows={5}
        value=""
      />
    </li>
  </ul>
  <button
    className="sc-ifAKCX hbBAat sc-iAyFgw koyBbc"
    disabled={true}
    type="submit"
  >
    Post
  </button>
</form>
`;
